### ISSUE TITLE
ttnn.linear produces incorrect tensor topology metadata for 2D mesh matmul

### ISSUE THREAD

**Issue by gwangTT** [tenstorrent/tt-metal#35362]
## Summary

`ttnn.linear` produces incorrect `tensor_topology()` metadata when operating on tensors sharded across a 2D mesh (e.g., Galaxy/TG topology). The output topology incorrectly claims both mesh axes shard the same tensor dimension.

## Reproduction

**Environment:** 2D mesh device (e.g., 4x8 or 8x4 Galaxy configuration)

```python
import torch
import ttnn

mesh_device = ...  # 2D mesh, e.g., shape (4, 8)
cluster_shape = list(mesh_device.shape)

dim, hidden_dim, seq_len = 4096, 14336, 32

# Input: shape [1, 1, 32, 4096], sharded on dim 3 across mesh axis 1
torch_input = torch.randn(1, 1, seq_len, dim, dtype=torch.bfloat16)
tt_input = ttnn.from_torch(
    torch_input,
    device=mesh_device,
    mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, dims=(None, 3), mesh_shape=cluster_shape),
    dtype=ttnn.bfloat16,
    memory_config=ttnn.DRAM_MEMORY_CONFIG,
    layout=ttnn.TILE_LAYOUT,
)

# Weight: shape [4096, 14336], sharded on both dims across both mesh axes
torch_weight = torch.randn(dim, hidden_dim, dtype=torch.bfloat16)
tt_weight = ttnn.from_torch(
    torch_weight,
    device=mesh_device,
    mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, dims=(-1, -2), mesh_shape=cluster_shape),
    dtype=ttnn.bfloat16,
    memory_config=ttnn.DRAM_MEMORY_CONFIG,
    layout=ttnn.TILE_LAYOUT,
)

# Run linear
tt_output = ttnn.linear(tt_input, tt_weight)
output_placements = list(tt_output.tensor_topology().placements())

print(output_placements)
# BUG: Shows [PlacementShard(-1), PlacementShard(3)]
# Both axes claim to shard the same dimension (dim 3 == dim -1 for a 4D tensor)!
```

## Expected Behavior

After `x @ w` where:
- `x` has topology `[Replicated, Shard(3)]`  
- `w` has topology `[Shard(-1), Shard(-2)]`

The output topology should be something like:
- `[Shard(3), PartialSum]` or
- `[Replicated, Shard(3)]` (after reduction on axis 0)

The key constraint: **both mesh axes cannot shard the same tensor dimension**.

## Actual Behavior

Output topology: `[PlacementShard(-1), PlacementShard(3)]`

For a 4D tensor, `dim=-1` normalizes to `dim=3`, so both mesh axes claim to shard dimension 3. This is physically impossible and breaks downstream operations that rely on topology metadata (e.g., `to_torch_auto_compose`, reduce-scatter, etc.).

## Impact

- `to_torch` with automatic mesh composition fails or produces incorrect results
- Users must manually specify `MeshComposerConfig` with the correct topology as a workaround
- Affects all 2D mesh matmul workloads (MLP, attention projections on Galaxy/TG)

## Workaround

Manually specify the correct topology when converting back to torch:

```python
# Known correct topology after all-reduce on axis 0
expected_composer_cfg = ttnn.MeshComposerConfig(
    dims=[0, 3],  # axis 0: replicated (dim ignored), axis 1: shard on dim 3
    mesh_shape_override=ttnn.MeshShape([1, cluster_shape[1]]),
)
mesh_composer = ttnn.create_mesh_composer(mesh_device, expected_composer_cfg)
tt_output_torch = ttnn.to_torch(tt_output, mesh_composer=mesh_composer)
```

## Test Reference

See `models/common/tests/modules/mlp/test_mlp_2d.py`:
- `test_ttnn_linear_2d_mesh_topology_bug` - xfail test documenting this bug
- `ttnn_linear_2d_mesh_has_topology_bug` fixture - minimal reproduction

**Comment by gwangTT** [comment]
Please take a look and re-assign, @ayerofieiev-tt @cmaryanTT. Thanks!

**Comment by bbradelTT** [comment]
Sonnet 4.5 in Cursor says this is due to something in ttnn/api/ttnn/device_operation.hpp. @ayerofieiev-tt could someone from your team look into this?

AI Output:

Why Does This Incorrect Behavior Occur?
The bug occurs due to missing dimension normalization in the default topology imputation logic. Here's the complete explanation:
Root Cause
The issue is in get_output_placements_and_shape function in ttnn/api/ttnn/device_operation.hpp (lines 365-462).
When you create tensors with ShardTensor2dMesh using negative dimensions (e.g., dims=(-1, -2)), these negative dimensions are stored directly in the PlacementShard without normalization. The topology imputation logic uses a map to track which tensor dimensions are already sharded:
std::unordered_map<int, int> shard_dim_to_distribution_dim;  // Line 393
The bug is at line 417 where it checks:
if (!shard_dim_to_distribution_dim.contains(new_shard_placement.dim))
The dimension is used as-is without normalization! So dim=3 and dim=-1 are treated as different map keys, even though for a 4D tensor they refer to the same dimension.
Step-by-Step Trace
With your reproduction example:
Input: [1, 1, 32, 4096] with [Replicate, Shard(3)]
Map stores: {3 → mesh_axis_1}
Weight: [4096, 14336] with [Shard(-1), Shard(-2)]
When broadcast to 4D: dim -1 = dim 3, dim -2 = dim 2
But placements still stored as Shard(-1) and Shard(-2)
Topology imputation processes weight:
   // Check if -1 is already in map   shard_dim_to_distribution_dim.contains(-1)  // FALSE! Only has key '3'      // Incorrectly allows duplicate sharding   shard_dim_to_distribution_dim[-1] = 0   result_placements[0] = Shard(-1)  // Which is dim 3!
Result: [Shard(-1), Shard(3)] where both -1 and 3 refer to dimension 3 in a 4D tensor!
Secondary Bug
There's also a bug at lines 439-446. When duplicate sharding IS detected, the code logs "replaced with Replicate" but never actually performs the replacement:
} else if (shard_dim_to_distribution_dim.at(new_shard_placement.dim) != i && kEnableLogging) {    log_warning(..., "replaced with Replicate", ...);    // ❌ MISSING: output_placement = Replicate{}}
Why It Matters
Since ttnn.linear doesn't implement custom compute_output_topologies, it falls back to this broken default imputation. The incorrect topology metadata can cause:
Downstream operations to fail
Incorrect distributed computation
Silent data corruption
The Fix
Two changes needed in ttnn/api/ttnn/device_operation.hpp:
Normalize dimensions before map lookup (around line 417):
auto new_shard_placement = std::get<...>(tensor_placements[i]);// Normalize negative dimensionsint normalized_dim = new_shard_placement.dim;if (normalized_dim < 0) {    int output_rank = /* infer from tensor shape */;    normalized_dim = output_rank + normalized_dim;}// Use normalized_dim for all comparisonsif (!shard_dim_to_distribution_dim.contains(normalized_dim)) { ... }
Actually replace duplicates with Replicate (around line 439):
} else if (shard_dim_to_distribution_dim.at(normalized_dim) != i) {    if (kEnableLogging) {        log_warning(...);    }    output_placement = tt::tt_metal::distributed::MeshMapperConfig::Replicate{};  // FIX}
I've created two files documenting this:
TOPOLOGY_BUG_EXPLANATION.md - Complete technical analysis
reproduce_topology_bug.py - Demonstration script showing the issue
This is a fundamental bug in how TTNN handles distributed tensor topology imputation when negative dimensions are involved!

[reproduce_topology_bug.py](https://github.com/user-attachments/files/24473138/reproduce_topology_bug.py)
[TOPOLOGY_BUG_EXPLANATION.md](https://github.com/user-attachments/files/24473139/TOPOLOGY_BUG_EXPLANATION.md)
